scapy/layers/roce.py

```diff
@@ -0,0 +1,199 @@
+    opcode('UC', 'RDMA_WRITE_LAST'),
+    opcode('UC', 'RDMA_WRITE_LAST_WITH_IMMEDIATE'),
+    opcode('UC', 'RDMA_WRITE_ONLY'),
+    opcode('UC', 'RDMA_WRITE_ONLY_WITH_IMMEDIATE'),
+
+    opcode('RD', 'SEND_FIRST'),
+    opcode('RD', 'SEND_MIDDLE'),
+    opcode('RD', 'SEND_LAST'),
+    opcode('RD', 'SEND_LAST_WITH_IMMEDIATE'),
+    opcode('RD', 'SEND_ONLY'),
+    opcode('RD', 'SEND_ONLY_WITH_IMMEDIATE'),
+    opcode('RD', 'RDMA_WRITE_FIRST'),
+    opcode('RD', 'RDMA_WRITE_MIDDLE'),
+    opcode('RD', 'RDMA_WRITE_LAST'),
+    opcode('RD', 'RDMA_WRITE_LAST_WITH_IMMEDIATE'),
+    opcode('RD', 'RDMA_WRITE_ONLY'),
+    opcode('RD', 'RDMA_WRITE_ONLY_WITH_IMMEDIATE'),
+    opcode('RD', 'RDMA_READ_REQUEST'),
+    opcode('RD', 'RDMA_READ_RESPONSE_FIRST'),
+    opcode('RD', 'RDMA_READ_RESPONSE_MIDDLE'),
+    opcode('RD', 'RDMA_READ_RESPONSE_LAST'),
+    opcode('RD', 'RDMA_READ_RESPONSE_ONLY'),
+    opcode('RD', 'ACKNOWLEDGE'),
+    opcode('RD', 'ATOMIC_ACKNOWLEDGE'),
+    opcode('RD', 'COMPARE_SWAP'),
+    opcode('RD', 'FETCH_ADD'),
+
+    opcode('UD', 'SEND_ONLY'),
+    opcode('UD', 'SEND_ONLY_WITH_IMMEDIATE'),
+
+    (CNP_OPCODE, 'CNP'),
+])
+
+
+class BTH(Packet):
+    name = "BTH"
+    fields_desc = [
+        ByteEnumField("opcode", 0, _bth_opcodes),
+        BitField("solicited", 0, 1),
+        BitField("migreq", 0, 1),
+        BitField("padcount", 0, 2),
+        BitField("version", 0, 4),
+        XShortField("pkey", 0xffff),
+        BitField("fecn", 0, 1),
+        BitField("becn", 0, 1),
+        BitField("resv6", 0, 6),
+        BitField("dqpn", 0, 24),
+        BitField("ackreq", 0, 1),
+        BitField("resv7", 0, 7),
+        BitField("psn", 0, 24),
+
+        FCSField("icrc", None, fmt="!I")]
+
+    @staticmethod
+    def pack_icrc(icrc):
+        return struct.pack("!I", icrc & 0xffffffff)[::-1]
+
+    def compute_icrc(self, p):
+        udp = self.underlayer
+        if udp is None:
+            warning("No underlayer to compute checksum. Leaving null.")
+            return self.pack_icrc(0)
+        ip = udp.underlayer
+        if isinstance(ip, IP):
+            # pseudo-LRH / IP / UDP / BTH / payload
+            pshdr = Raw(b'\xff' * 8) / ip.copy()
+            pshdr.chksum = 0xffff
+            pshdr.ttl = 0xff
+            pshdr.tos = 0xff
+            pshdr[UDP].chksum = 0xffff
+            pshdr[BTH].fecn = 1
+            pshdr[BTH].becn = 1
+            pshdr[BTH].resv6 = 0xff
+            bth = pshdr[BTH].self_build()
+            payload = raw(pshdr[BTH].payload)
+            # add ICRC placeholder just to get the right IP.totlen and
+            # UDP.length
+            icrc_placeholder = b'\xff\xff\xff\xff'
+            pshdr[UDP].payload = Raw(bth + payload + icrc_placeholder)
+            icrc = crc32(raw(pshdr)[:-4]) & 0xffffffff
+            return self.pack_icrc(icrc)
+        else:
+            # TODO support IPv6
+            warning("No IP underlayer to compute checksum. Leaving null.")
```

Comment:
You should display a warning saying that the `underlayer` protocol is not supported.

Reply:
Done. I tried adding the underlayer name as well to make the warning clear.